bahmni_atom_feed/models/atom_event_worker.py

```diff
@@ -0,0 +1,91 @@
+# -*- coding: utf-8 -*-
+from odoo import models, fields, api
+import logging
+import json
+import uuid
+
+STATE_CODE_PREFIX = 'UNKNOWN-'
+_logger = logging.getLogger(__name__)
+
+
+class AtomEventWorker(models.Model):
+    _name = 'atom.event.worker'
+    _auto = False
+
+    @api.model
+    def process_event(self, vals):
+        '''Method getting triggered from Bahmni side'''
+        category = vals.get("category")
+        try:
+            if category == "create.customer":
+                self._create_or_update_customer(vals)
+            return {'success': True}    
+        except Exception as err:
+            _logger.info("\n Processing event threw error: %s", err)
+            raise
+
+
+    @api.model
+    def _create_or_update_customer(self, vals):
+        patient_ref = vals.get("ref")
+        customer_vals = self._get_customer_vals(vals)
+        # removing null values, as while updating null values in write method will remove old values
+
+        for rec in customer_vals.keys():
+            if not customer_vals[rec]:
+                del customer_vals[rec]
+        existing_customer = self.env['res.partner'].sudo().search([('ref', '=', patient_ref)])
+        if existing_customer:
+            existing_customer.write(customer_vals)
+            self._create_or_update_person_attributes(existing_customer.id,vals)
+        else:
+            customer = self.env['res.partner'].sudo().create(customer_vals)
+            self._create_or_update_person_attributes(customer.id,vals)
+
+    @api.model
+    def _get_address_details(self, address):
+        res = {}
+        if address.get('address1'):
+            res.update({'street': address['address1']})
+        if address.get('address2'):
+            res.update({'street2': address['address2']})        
+
+        return res
+
+
+    def _get_customer_vals(self, vals):
+        res = {}
+        res.update({'ref': vals.get('ref'),
+                    'name': vals.get('name'),
+                    'local_name': vals.get('local_name'),
+                    'uuid': vals.get('uuid'),
+                    })
+        address_data = vals.get('preferredAddress')
+        # get validated address details
+        address_details = self._get_address_details(json.loads(address_data))
+        # update address details
+        res.update(address_details)
+        # update other details : for now there is only scope of updating contact.
+        if vals.get('primaryContact'):
+            res.update({'phone': vals['primaryContact']})
+        return res
+        
+    def _create_or_update_person_attributes(self, cust_id, vals):
+        attributes = json.loads(vals.get("attributes", "{}"))
+        attributes_list = ['landHolding', 'familyNameLocal', 'cluster', 'education', 'occupation','caste', 'isUrban', 'primaryRelative',
+                'familyIncome','secondaryContact','middleNameLocal','RationCard', 'givenNameLocal', 'distanceFromCenter',
```

Comment:
The list of attributes seems to be hardcoded here. This is not the case in BAHMNI. The attributes can be configured by a hospital to their own custom list. The code needs to treat this as generic key/value pairs.

Reply:
Hardcoded values have been eliminated, and the loading of additional info tab is now based on the key/value pairs from the API attributes data."